azurelinuxagent/common/rdma.py

```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
+            return (key_0, value_0)
+
+        return (None, None)
+
+    def provision_sriov_multiple_ib(self, value) :
+
+        mac_ip_array = []
+
+        values = value.split("|")
+        num_ips = len(values) - 1
+        # values[0] tells how many IPs. Format - NUMPAIRS:<number>
+        match = re.match(r"NUMPAIRS:(\d+)", values[0])
+        if match:
+            num = int(match.groups(0)[0])
+            if num != num_ips:
+                logger.error("RDMA: multiple IPs reported num={0} actual number of IPs={1}".format(num, num_ips))
+                return
+        else:
+            logger.error("RDMA: failed to find number of IP addresses in {0}".format(values[0]))
+            return
+
+        for i in range(1, num_ips+1):
+            # each MAC/IP entry is of format <MAC>:<IP>
+            match = re.match(r"([^:]+):(\d+\.\d+\.\d+\.\d+)", values[i])
+            if match:
+                mac_addr = match.groups(0)[0]
+                ipv4_addr = match.groups(0)[1]
+                mac_ip_array.append((mac_addr, ipv4_addr))
+            else:
+                logger.error("RDMA: failed to find MAC/IP address in {0}".format(values[i]))
+                return
+
+        # try to assign all MAC/IP addresses to IB interfaces
+        # retry for up to 60 times, with 1 seconds delay between each
+        retry = 60
+        while retry > 0:
+            count = self.update_iboip_interfaces(mac_ip_array)
+            if count == len(mac_ip_array):
+                return
+
+            time.sleep(1)
+            retry -= 1
+
+        logger.error("RDMA: failed to set all IP over IB addresses")
+
+    # Assign addresses to all IP over IB interfaces specified in mac_ip_array
+    # Return the number of IP addresses successfully assigned
+
+    def update_iboip_interfaces(self, mac_ip_array):
+
+        net_dir = "/sys/class/net"
+        nics = os.listdir(net_dir)
+        count = 0
+
+        for nic in nics:
+            # look for IBoIP interface of format ibXXX
+            if not re.match(r"ib\d+", nic):
+                continue
+            address_file = open(net_dir + "/" + nic + "/address")
+            mac_addr = address_file.read()
+            address_file.close()
+
+            if not mac_addr:
+                logger.error("RDMA: can't read address for device {0}".format(nic))
+                continue
+
+            mac_addr = mac_addr.upper()
+
+            match = re.match(r".+(\w\w):(\w\w):(\w\w):\w\w:\w\w:(\w\w):(\w\w):(\w\w)\n", mac_addr)
+            if not match:
+                logger.error("RDMA: failed to parse address for device {0} address {1}".format(nic, mac_addr))
+                continue
+
+            # format an MAC address without :
+            mac_addr = ""
+            mac_addr = mac_addr.join(match.groups(0))
+
+            for mac_ip in mac_ip_array:
+                if mac_ip[0] == mac_addr:
+                    ret = shellutil.run("ip addr add {0}/16 dev {1}".format(mac_ip[1], nic))
```

Comment:
We're deprecating the use of `shellutil.run()` function. Please use `shellutil.run_command()` here instead

Reply:
shellutil.run() will not raise an exception on error. Is there a way to not to raise an exception with shellutil.run_command() and return error code?

Reply:
you can just catch it and capture the return code like that. Check out the implementation here for an example - https://github.com/Azure/WALinuxAgent/blob/5e97cf23b53d53c95247e06ff3c93a85029271a3/azurelinuxagent/common/osutil/redhat.py#L111

Basically something like - 
```python
        ret_code = 0
        try:
            return shellutil.run_command(cmd, log_error=True)
        except shellutil.CommandError as error: 
            ret_code = error.returncode


```